all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error>;
+
+  /// Tries to immediately reserve a slot, returning true if a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  fn try_reserve_slot(&self) -> Result<bool, anyhow::Error>;
+
+  /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
+  /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
+  /// then the slot can actually be used to work on the task.
+  /// 
+  /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
+  /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
+  fn mark_slot_used(&self, info: &Self::SlotKind::Info);
+
+  /// Frees a slot.
+  fn release_slot(&self, info: &Self::SlotKind::Info);
+  
+  /// If this implementation knows how many slots are available at any moment, it should return that here.
+  fn available_slots(&self) -> Option<usize>;
+}
+
+// These functions would be provided to the user as top level statics they can access to get information about how
+// all outstanding slots are currently being used. Ostensibly they could've stored this in their own implementation
+// too, but this is to make it convenient for them to answer questions like "How many slots are being used by 
+// activities of type X right now?". 
+//
+// These could potentially be provided as parameters to the trait methods - but at least in the case of the 
+// async/blocking reserve, the information may change over time, so it may be best to have these be explicit calls.
+pub fn used_workflow_slots_info() -> WorkflowSlotsInfo { /* ... */ }
+pub fn used_activity_slots_info() -> ActivitySlotsInfo { /* ... */ }
+pub fn used_local_activity_slots_info() -> LocalActivitySlotsInfo { /* ... */ }
```

Comment:
I think this stuff can come later. The slot supplier impl can already keep general counts (and if they need to differentiate local from non-local activities, we should provide the contextual information at reservation time). As for count per workflow/activity type, I think we can expand to that later.

Reply:
Because it blurs the line between who holds what info. If you want to provide contextual info on invocation of reserve/release, ok, but to have a slot supplier and our system maintain counts and such seems like there's a bit of redudancy and duplication. If given the option between potentially useful and simple initial interface, I think we should err on the side of the latter initially.

Reply:
Not that easy. We are now going to clone/serialize every slot in use across the FFI boundary for every lang side that calls it. And to be thread-safe, even if not across a lang boundary, we're probably going to have to copy a snapshot of this info upon request. I do see how it is useful info, I'm just worried if we're assuming people need more data than they need today. Even with middleground, if you got rid of the vectors of slot infos, you'd just have the same counts a user could easily keep track of.

Reply:
>We shouldn't need to track it outside of the slot supplier.

We must: "used slots" ~= "in-flight tasks". In flight tasks must be tracked for obvious reasons. These are stored today as your maps of `workflow id -> state about current task/workflow/etc` and `task token -> state about activity task` etc.

These methods would provide a read-only view into some lens on that tracked data.

Reply:
Sure in flight tasks must be tracked by us, I just question whether the user needs to be able to access them to make their decisions. And of course there is the cost of copying that info upon access (or otherwise synchronizing access).

Also arguably this is general purpose per-worker stat info request unrelated to this project unless we think users of this project need this general purpose info. I wonder if we should have a concept of "get status" on workers and pass the worker that is asking as part of each call.

Reply:
So Go and Java currently have no such map or info explicitly tracked by the SDK. I wouldn't want to provide that info from the SDK unless there was a strong user ask and there was no other way to implement the ask. I agree with Chad that this seems unrelated to auto tuning and maybe should be a separate proposal

Reply:
It must exist at least partially in some form somewhere, otherwise how are the workflow/activity contexts supplying the various bits of information they do?

I agree that this could be some separate generic worker info call.

So, anyway if this is hard to provide in Go/Java (I can't imagine it is really, since I'm not thinking that a huge amount of stuff is actually going to end up in these structs, activity/wf type is the primary initial motivation). 

That said, that can also argue the other way that it's easy for the implementer to track themselves. So, that in mind sure, I'll take this out and we can bring it back later if it becomes important.

Reply:
>It must exist at least partially in some form somewhere, otherwise how are the workflow/activity contexts supplying the various bits of information they do?

It doesn't exist in any centralized map